Pinta.Effects/Effects/OutlineObjectEffect.cs

```diff
@@ -0,0 +1,192 @@
+
+		// First pass
+		// Clean up dest, then collect all border pixels
+		Parallel.For (top, bottom + 1, new ParallelOptions { MaxDegreeOfParallelism = threads }, y => {
+			var srcData = src.GetReadOnlyPixelData ();
+
+			// reset dest to src
+			// Removing this causes preview to not update to lower radius levels
+			var srcRow = srcData.Slice (y * srcWidth, srcWidth);
+			var dstRow = dest.GetPixelData ().Slice (y * srcWidth, srcWidth);
+			for (int x = left; x <= right; x++)
+				dstRow[x].Bgra = srcRow[x].Bgra;
+
+			// Produces different behaviour at radius == 0 and radius == 1
+			// When radius == 0, only consider direct border pixels
+			// When radius == 1, consider border pixels on diagonal
+			Span<PointI> pixels = stackalloc PointI[8];
+			// Collect a list of pixels that surround the object (border pixels)
+			for (int x = left; x <= right; x++) {
+				PointI potentialBorderPixel = new (x, y);
+				if (Data.OutlineBorder && (x == 0 || x == srcWidth - 1 || y == 0 || y == srcHeight - 1)) {
+					borderPixels.Add (potentialBorderPixel);
+				} else if (src.GetColorBgra (srcData, srcWidth, potentialBorderPixel).A <= Data.Tolerance) {
+					// Test pixel above, below, left, & right
+					pixels[0] = new (x - 1, y);
+					pixels[1] = new (x + 1, y);
+					pixels[2] = new (x, y - 1);
+					pixels[3] = new (x, y + 1);
+					int pixelCount = 4;
+					if (radius == 1) {
+						// if radius == 1, also test pixels on diagonals
+						pixels[4] = new (x - 1, y - 1);
+						pixels[5] = new (x - 1, y + 1);
+						pixels[6] = new (x + 1, y - 1);
+						pixels[7] = new (x + 1, y + 1);
+						pixelCount = 8;
+					}
+
+					for (int i = 0; i < pixelCount; i++) {
+						var px = pixels[i].X;
+						var py = pixels[i].Y;
+						if (px < 0 || px >= srcWidth || py < 0 || py >= srcHeight)
+							continue;
+						if (src.GetColorBgra (srcData, srcWidth, new PointI (px, py)).A > Data.Tolerance) {
+							borderPixels.Add (potentialBorderPixel);
+							// Remove comments below to draw border pixels
+							// You will also have to comment out the 2nd pass because it will overwrite this
+							//int pos = srcWidth * y + x;
+							//borderData[pos].Bgra = 0;
+							//borderData[pos].A = 255;
+
+							break;
+						}
+					}
+
+
+				}
+			}
+		});
+
+
+		// Second pass
+		// Generate outline and blend to dest
+		Parallel.For (top, bottom + 1, new ParallelOptions { MaxDegreeOfParallelism = threads }, y => {
+			// otherwise produces nothing at radius == 0
+			if (radius == 0)
+				radius = 1;
+			var relevantBorderPixels = borderPixels.Where (borderPixel => borderPixel.Y > y - radius && borderPixel.Y < y + radius).ToArray ();
+			var destRow = dest.GetPixelData ().Slice (y * srcWidth, srcWidth);
+
+			for (int x = left; x <= right; x++) {
+				byte highestAlpha = 0;
+
+				// optimization: no change if destination has max alpha already
+				if (destRow[x].A == 255)
+					continue;
+
+				if (Data.FillObjectBackground && destRow[x].A >= Data.Tolerance)
```

Comment:
I'm not entirely sure if I'm understanding the use case for the Fill Object Background toggle, but it might be more clear to label it as making the object opaque since it's not filling with a new colour?

Reply:
To show, 1st is pic before, 2nd is without, 3rd is with fill object background

![image](https://github.com/user-attachments/assets/e29955fa-36c3-4c83-992d-3fb8a660aee2)
![image](https://github.com/user-attachments/assets/ef42fec8-c211-4b53-96c4-7c0221ea3001)
![image](https://github.com/user-attachments/assets/bbc264df-168c-461d-a3d4-d6e6bc5eada4)

I'm also not too sure what to name this, but I decided "fill object background" was most fitting.

Reply:
Thanks, yeah that seems like a reasonable label then 👍